[GreyNoise] Add Integration Package

niraj-crest opened the pull request:
- Enhancement

## What does this PR do?

- Added 1 data stream (ip).
- Added data collection logic for the data streams.
- Added the ingest pipeline for the data streams.
- Mapped fields according to the ECS schema and added Fields metadata in the appropriate yaml files.
- Added dashboards and visualizations.
- Added system test cases for the data stream.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

- Clone integrations repo.
- Install elastic-package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/ti_greynoise) directory.
- Run the following command to run tests. elastic-package test


<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

kcreddy commented:
/test

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/25697)
* Commit: 257d5f6e60d4f245fbb803fecfecee1bc1f5572b

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25697","number":25697,"commit":"257d5f6e60d4f245fbb803fecfecee1bc1f5572b"}],"number":25697}
buildkite-pr-comment-->

efd6 commented on packages/ti_greynoise/_dev/deploy/docker/files/config.yml:
```diff
@@ -0,0 +1,196 @@
+                  "complete": true
+              }
+          }
+          `}}
```
Add final new line.

efd6 commented on packages/ti_greynoise/data_stream/ip/_dev/test/pipeline/test-ip.log:
```diff
@@ -0,0 +1 @@
+{ "ip": "1.128.0.0", "business_service_intelligence": { "found": true, "category": "public_dns", "name": "Google Public DNS", "description": "Google's global domain name system (DNS) resolution service.", "explanation": "Public DNS services are used as alternatives to ISP's name servers. You may see devices on your network communicating with Google Public DNS over port 53/TCP or 53/UDP to resolve ...
```
Add final new line.

efd6 commented on packages/ti_greynoise/_dev/deploy/docker/files/config.yml:
```diff
@@ -0,0 +1,196 @@
+rules:
```
Please make the system tests do an initial, a continuation and a termination in the paginations. This would require an additional rule between the two that are here.

efd6 commented on packages/ti_greynoise/data_stream/ip/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,84 @@
+          "error": {
+            "code": string(resp.StatusCode),
+            "id": string(resp.Status),
+            "message": ":" + (
```
```suggestion
            "message": "GET " + state.url.trim_right("/") + "/v3/gnql: " + (
```

efd6 commented on packages/ti_greynoise/data_stream/ip/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,84 @@
+  state.with(
+    request(
+      "GET",
+      state.url.trim_right("/") + "/v3/gnql" + "?" + {
```
```suggestion
      state.url.trim_right("/") + "/v3/gnql?" + {
```

efd6 commented on packages/ti_greynoise/data_stream/ip/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,84 @@
+    request(
+      "GET",
+      state.url.trim_right("/") + "/v3/gnql" + "?" + {
+      ?"scroll": has(state.scroll) && state.scroll != null ? optional.of([state.scroll]) : optional.none(),
```
```suggestion
      ?"scroll": has(state.?next.scroll) ?
        optional.of([state.next.scroll])
      :
        optional.none(),
```

efd6 commented on packages/ti_greynoise/data_stream/ip/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,84 @@
+      state.url.trim_right("/") + "/v3/gnql" + "?" + {
+      ?"scroll": has(state.scroll) && state.scroll != null ? optional.of([state.scroll]) : optional.none(),
+      "query": ["last_seen:1d" + (has(state.query) && state.query != null ? " " + state.query.trim_right(" ") : "")],
+      "size": [(has(state.page_size) && state.page_size != null && int(state.page_size) <= 5000 ? string(state.page_size) : "5000")],
```
```suggestion
      "size": [int(state.page_size) <= 5000 ? string(state.page_size) : "5000"],
```

efd6 commented on packages/ti_greynoise/data_stream/ip/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,84 @@
+            [{"message": "retry"}]
+        ),
+        "want_more": has(body.?request_metadata.complete) && body.request_metadata.complete == false,
+        "scroll": has(body.?request_metadata.scroll) ? body.request_metadata.scroll : null
```
```suggestion
        "next": {"scroll": body.?request_metadata.scroll},
```
and update all the references to `state.scroll` with `state.next.scroll`

efd6 commented on packages/ti_greynoise/data_stream/ip/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,84 @@
+              ),
+          },
+        },
+        "scroll": null,
```
```suggestion
        "next": {},
```

efd6 commented on packages/ti_greynoise/data_stream/ip/manifest.yml:
```diff
@@ -0,0 +1,73 @@
+        required: true
+        show_user: false
+        default: 5m
+      - name: enable_request_tracer
+        type: bool
+        title: Enable request tracing
+        multi: false
+        required: false
+        show_user: false
+        description: The request tracer logs requests and responses to the agent's local file-system for debugging configurations. Enabling this request tracing compromises security and should only be used for debugging. See [documentation](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-cel.html#_resource_tracer_filename) for details.
```
Add a `default: false`.

efd6 commented on packages/ti_greynoise/_dev/build/docs/README.md:
```diff
@@ -0,0 +1,124 @@
+- Only one Elastic Agent can be installed per host.
+- Elastic Agent is required to stream data through the REST API and ship it to Elastic, where events are processed via the integration's ingest pipelines.
+
+### Installing and Managing Elastic Agent
+
+You have several options for installing and managing Elastic Agent:
+
+#### Install a Fleet-managed Elastic Agent (Recommended)
+
+With this approach, you install Elastic Agent and use Fleet in Kibana to define, configure, and manage your agents centrally. We recommend using Fleet management because it significantly simplifies agent management and upgrades.
+
+#### Install Elastic Agent in Standalone Mode (Advanced Users)
+
+With this approach, you install Elastic Agent and manually configure it locally on the system where it's installed. You are responsible for managing and upgrading the agents. This approach is reserved for advanced users only.
+
+#### Install Elastic Agent in a Containerized Environment
+
+You can run Elastic Agent inside a container, either with Fleet Server or standalone. Docker images for all versions of Elastic Agent are available from the Elastic Docker registry, and we provide deployment manifests for running on Kubernetes.
+
+For minimum requirements and more information, refer to the Elastic Agent [installation guide](https://www.elastic.co/guide/en/fleet/current/elastic-agent-installation.html).
```
See changes [here](https://github.com/elastic/integrations/pull/13573).

efd6 commented on packages/ti_greynoise/data_stream/ip/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,422 @@
+      target_field: threat.indicator.as.number
+      pattern: 'AS'
+      replacement: ''
+      if: ctx.greynoise?.ip?.internet_scanner_intelligence?.metadata?.asn != null && ctx.greynoise?.ip?.internet_scanner_intelligence?.metadata?.asn != ''
```
```suggestion
      if: ctx.greynoise?.ip?.internet_scanner_intelligence?.metadata?.asn != null && ctx.greynoise.ip.internet_scanner_intelligence.metadata.asn != ''
```

efd6 commented on packages/ti_greynoise/data_stream/ip/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,422 @@
+      field: threat.indicator.reference
+      tag: set_threat_indicator_reference_from_ip_internet_scanner_intelligence_ip
+      value: 'https://www.greynoise.io/ip/{{{threat.indicator.ip}}}'
+      if: ctx.threat?.indicator?.ip != null && ctx.threat?.indicator?.ip != ''
```
```suggestion
      if: ctx.threat?.indicator?.ip != null && ctx.threat.indicator.ip != ''
```

efd6 commented on packages/ti_greynoise/data_stream/ip/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,422 @@
+      field: threat.indicator.description
+      tag: set_threat_indicator_description
+      value: '{{{threat.indicator.ip}}} IP has been observed mass scanning the internet by GreyNoise with a classification of {{{greynoise.ip.internet_scanner_intelligence.classification}}}'
+      if: ctx.threat?.indicator?.ip != null && ctx.threat?.indicator?.ip != '' && ctx.greynoise?.ip?.internet_scanner_intelligence?.classification != null && ctx.greynoise?.ip?.internet_scanner_intelligence?.classification != ''
```
```suggestion
      if: >-
        ctx.threat?.indicator?.ip != null &&
        ctx.threat.indicator.ip != '' &&
        ctx.greynoise?.ip?.internet_scanner_intelligence?.classification != null &&
        ctx.greynoise.ip.internet_scanner_intelligence.classification != ''
```

kcreddy commented on packages/ti_greynoise/kibana/security_rule/018087be-76a7-4106-b41a-caa9f8d9184a.json:
```diff
@@ -0,0 +1,105 @@
+{
```
Curious why we be needing this rule when we have existing [Threat Intel IP rule](https://www.elastic.co/docs/reference/security/prebuilt-rules/rules/threat_intel/threat_intel_indicator_match_address)

niraj-crest replied:
Primarily to set `Indicator prefix override: greynoise.ip`, which will allow us to build enrichment dashboards once we're ready to integrate the transform pipeline.

kcreddy replied:
> this may not align with Elastic's recommended best practices.

Alright. btw can you share link(s) to best practices you are following right now?

kcreddy replied:
Well it depends case by case basis, but we did add extended ECS fields before, commonly inside `process`, `vulnerability`, `host`. 
Although what I meant to say was not copy everything from `greynoise.ip` into `threat.indicator`, but some important fields that you may later use for dashboards.

niraj-crest replied:
Understood, @kcreddy 

We’re planning to include the following fields in the detected threat event. These fields provide valuable context for users, especially when creating custom visualizations based on detected threats.

For reference, here are the related documents from GreyNoise:

- [Triage](https://github.com/elastic/integrations/pull/link)

- [Business Services](https://github.com/elastic/integrations/pull/link)

**Planned Fields**

- internet_scanner_intelligence.actor
- internet_scanner_intelligence.bot
- internet_scanner_intelligence.classification
- ip
- internet_scanner_intelligence.last_seen_timestamp
- internet_scanner_intelligence.last_seen
- internet_scanner_intelligence.metadata.asn
- internet_scanner_intelligence.metadata.category
- internet_scanner_intelligence.metadata.source_city
- internet_scanner_intelligence.metadata.mobile
- internet_scanner_intelligence.metadata.organization
- internet_scanner_intelligence.metadata.source_country
- internet_scanner_intelligence.metadata.source_country_code
- internet_scanner_intelligence.metadata.rdns
- internet_scanner_intelligence.metadata.region
- internet_scanner_intelligence.tor
- internet_scanner_intelligence.found
- internet_scanner_intelligence.spoofable
- internet_scanner_intelligence.tags
- internet_scanner_intelligence.vpn
- internet_scanner_intelligence.vpn_service
- business_service_intelligence.category
- business_service_intelligence.description
- business_service_intelligence.explanation
- business_service_intelligence.last_updated
- business_service_intelligence.name
- business_service_intelligence.reference
- business_service_intelligence.trust_level

I will add `ecs-threat-extended.yml` file to maintain the extended ECS fields.

kcreddy replied:
There seems to be too many @niraj-crest. Do you need all of them for the matches dashboard?

kcreddy replied:
It does seem like we have to add a lot of fields. I am beginning to feel uncertain if this approach is correct.
Ideally, I would've liked that Elastic Security Intelligence app handles the indicator matches dashboard because its where the Alerts also reside. That way we have everything (alerts + indicator matches) in one place. Instead of us trying to create this dashboard at integration level.

@jamiehynds, this is a similar discussion we had over DM which stemmed from https://github.com/elastic/integrations/pull/13189#discussion_r2028144454. Essentially if we want to create these additional dashboards about indicator matches (after receiving alerts), we are required to take out-of-norm step, either by creating a detection rule (a near duplicate of IP indicator rule, except for `threat_indicator_path = "threat.indicator"` is replaced by `threat_indicator_path = "greynoise.ip"`) or defining[ several custom fields](https://github.com/elastic/integrations/pull/13745#discussion_r2084293350) inside `threat.indicator` ECS mapping itself. Note that we haven't had these additional dashboards for existing Threat Intel integrations.
Let me know your thoughts if we really need these dashboards and if so which would be the preferred way.

niraj-crest replied:
Hey @kcreddy/@jamiehynds, 
Do we have any update on which approach to follow?

jamiehynds replied:
@niraj-crest customers (and I assume Greynoise) will expect that when they ingest indicators from Greynoise, that our out of the box threat intelligence dashboards and indicator match rules will automatically leverage Greynoise data. For that to happen, the data needs to be mapped to the correct ECS fields. (you can check the dashboards and indicator match rules to see exactly which fields are required). 

Custom fields are ok to use as part of an integration, in cases where there isn't a suitable field in ECS. These custom fields should be prefixed with `greynoise.*` and can be used for custom dashboards if required.

niraj-crest replied:
Thanks, we’ve reviewed the built-in threat intel rules and ensured that our GreyNoise pipeline includes the relevant ECS field mappings required by those rules.

In addition, we’ve implemented custom rules that use `kibana.alert.rule.tags: "GreyNoise"` and `threat_indicator_path: "greynoise.ip"`. These help the rule transform in this [Enhancement PR](https://github.com/elastic/integrations/pull/13786) detect alerts generated through indicator matches from GreyNoise data. The detected alerts are then used to populate our custom threat intel dashboards, which provide visibility into the IPs within the user's environment that matched GreyNoise indicators. This gives users actionable context and enhances threat intel visibility beyond the standard rules.

Please let me know if you have any other questions.

kcreddy replied:
> @niraj-crest customers (and I assume Greynoise) will expect that when they ingest indicators from Greynoise, that our out of the box threat intelligence dashboards and indicator match rules will automatically leverage Greynoise data.

When the users install/enable this new custom rule for Greynoise, they should uninstall/disable the standard Elastic IP-based rule. This is to avoid duplicate alerts coming from same indicator. This seems to be confusing for the users as this integration is distinctly doing certain things that other TI* integrations are not doing. 

> To address this, we propose implementing a custom detection rule that will populate the necessary data for these custom dashboards and provide more detailed insights into detected threats.

I still somehow believe that these additional insights should be supported out-of-the-box inside the Elastic `Intelligence` app, probably by creating `.internal.alert-ioc-matches*` indices, instead of integration creating its own rules / transform pipelines for every integration and data stream.

kcreddy replied:
> The default [Elastic IP-based rule](https://github.com/elastic/detection-rules/blob/main/rules/threat_intel/threat_intel_indicator_match_address.toml) uses a condition such as NOT labels.is_ioc_transform_source: "true". To prevent duplicate alerts, we’ve implemented the labels.is_transform_source flag. ensuring both rules operate independently without conflict.

It is good to know that duplicate alerts wont occur. 

But this is still deviating from original approach for all threat intel integration. In the future if we want to add any functionality to all threat intel transforms/destination indices, this custom approach makes it difficult.
I am just not in favour of creating custom threat intel rule(s) per integration in favour of providing additional value as it could become maintenance burden. But I am not going to block it.

@jamiehynds, let me know if you are okay with this approach. Will approve to unblock.

niraj-crest replied:
hello @jamiehynds  as discussed at here

> In addition, we’ve implemented custom rules that use `kibana.alert.rule.tags: "GreyNoise"` and `threat_indicator_path: "greynoise.ip"`. These help the rule transform in this [Enhancement PR](https://github.com/elastic/integrations/pull/13786) detect alerts generated through indicator matches from GreyNoise data. The detected alerts are then used to populate our custom threat intel dashboards, which provide visibility into the IPs within the user's environment that matched GreyNoise indicators. This gives users actionable context and enhances threat intel visibility beyond the standard rules.

We’ve added custom rules specifically designed to detect indicator match alerts from `GreyNoise` data. These alerts directly power our custom threat intel dashboards, helping deliver richer context and improved visibility to users.

The existing rule remains unaffected by this addition, as we’ve taken care to ensure both rules operate independently without conflict.

Appreciate your consideration, and hoping we can move forward with this approach.

kcreddy commented on packages/ti_greynoise/manifest.yml:
```diff
@@ -0,0 +1,90 @@
+type: integration
+categories:
+  - custom
+  - threat_intel
```
Add `- security`

kcreddy commented on packages/ti_greynoise/manifest.yml:
```diff
@@ -0,0 +1,90 @@
+description: Collect Threat Intelligence Indicators from GreyNoise using Elastic Agent, and perform enrichment on Elasticsearch by correlating Indicators of Compromise (IOCs).
+type: integration
+categories:
+  - custom
```
```suggestion
```

kcreddy commented on packages/ti_greynoise/manifest.yml:
```diff
@@ -0,0 +1,90 @@
+              #    -----END CERTIFICATE-----
+owner:
+  github: elastic/security-service-integrations
+  type: partner
```
We have been using `type: elastic` for integrations built by Crest. Can you please confirm if the `type` should really be `partner`?

kcreddy commented on packages/ti_greynoise/img/greynoise_threat_feed_overview_dashboard.png:
Add a saved search at the bottom if possible.

kcreddy commented on packages/ti_greynoise/img/greynoise_threat_feed_overview_dashboard.png:
Add screenshot to PR description.

efd6 commented on packages/ti_greynoise/data_stream/ip/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,87 @@
+            [{"message": "retry"}]
+        ),
+        "want_more": has(body.?request_metadata.complete) && body.request_metadata.complete == false,
+        "next": {"scroll": has(body.?request_metadata.scroll) ? body.request_metadata.scroll : null},
```
```suggestion
        "next": {?"scroll": body.?request_metadata.scroll},
```

niraj-crest commented:
Hello @kcreddy,
We’ve addressed all the review comments.
Kindly let us know if you have any additional suggestions. If everything looks good to you, we’d appreciate it if we could proceed with merging this PR.

efd6 commented:
/test

kcreddy reviewed (commented):
Thanks @niraj-crest for resolving the comments.
Except for https://github.com/elastic/integrations/pull/13745#discussion_r2135621706, the rest of the PR looks good.

niraj-crest commented:
Hello @kcreddy,
We have replied to comment can you please review it. https://github.com/elastic/integrations/pull/13745#discussion_r2139823637

jamiehynds commented:
@niraj-crest the last remaining blocker for this integration relates to shipping the Greynoise specific rules with the integration. I discussed with our detection rules team, and they recommend that we remove the rule(s) from this package, and instead, add the Greynoise rule(s) to our OOTB rules. We can also map the Greynoise integration to that rule as a required integration in our UI. 

Are you ok with removing the rule to allow us to merge the Greynoise integration. Then, raise an issue in our [detections repo](https://github.com/elastic/detection-rules) to go about adding the Greynoise rules to our OOTB rules?

niraj-crest commented:
Hello @jamiehynds,
As suggested, we’ve removed the rule from this PR and have raised an issue in the OOTB repository [elastic/detection-rules#4839](https://github.com/elastic/detection-rules/issues/4839). This should help us move forward and merge this PR.

Our second draft PR ([#13786](https://github.com/elastic/integrations/pull/13786)) will depend on the resolution of this issue. If possible, we’d appreciate it if this could be prioritized or alternatively, we’re happy to raise a PR directly in the detection rules repo if that helps move things along.

niraj-crest commented:
Hello @jamiehynds, @kcreddy,

Any updates on this Comment. https://github.com/elastic/integrations/pull/13745#issuecomment-2995016839
If everything looks good to you, could you please approve, and merge this PR?
Thank you!

kcreddy commented:
/test

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/27678)
* Commit: 30191e13045153540af060d82a6b11a273ef043c

### History
* :green_heart: [Build #27062](https://buildkite.com/elastic/integrations/builds/27062) succeeded 06d89c08ed0dd4144031c47ff829e6a443abec3c
* :green_heart: [Build #26898](https://buildkite.com/elastic/integrations/builds/26898) succeeded 839a9005512204a87f81e14fbdff27a94c77729c
* :green_heart: [Build #25697](https://buildkite.com/elastic/integrations/builds/25697) succeeded 257d5f6e60d4f245fbb803fecfecee1bc1f5572b

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/27678","number":27678,"commit":"30191e13045153540af060d82a6b11a273ef043c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/27062","number":27062,"commit":"06d89c08ed0dd4144031c47ff829e6a443abec3c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/26898","number":26898,"commit":"839a9005512204a87f81e14fbdff27a94c77729c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25697","number":25697,"commit":"257d5f6e60d4f245fbb803fecfecee1bc1f5572b"}],"number":27678}
buildkite-pr-comment-->
